src/ua_types_encoding_json.c

```diff
@@ -0,0 +1,5101 @@
+            return ret;
+    }
+    
+    if (src->hasAdditionalInfo) {
+        ret |= calcWriteKey(ctx, "AdditionalInfo", commaNeeded);
+        commaNeeded = UA_TRUE;
+        ret |= CALC_DIRECT(&src->additionalInfo, String);
+        if (ret != UA_STATUSCODE_GOOD)
+            return ret;
+    }
+
+    if (src->hasInnerStatusCode) {
+        ret |= calcWriteKey(ctx, "InnerStatusCode", commaNeeded);
+        commaNeeded = UA_TRUE;
+        ret |= CALC_DIRECT(&src->innerStatusCode, StatusCode);
+        if (ret != UA_STATUSCODE_GOOD)
+            return ret;
+    }
+
+    if (src->hasInnerDiagnosticInfo) {
+        ret |= calcWriteKey(ctx, "InnerDiagnosticInfo", commaNeeded);
+        /*Check recursion depth in encodeJsonInternal*/
+        ret |= calcJsonInternal(src->innerDiagnosticInfo, &UA_TYPES[UA_TYPES_DIAGNOSTICINFO], ctx, useReversible);
+        if (ret != UA_STATUSCODE_GOOD)
+            return ret;
+    }
+
+    ret |= ADDJSONCHAR(ObjEnd);
+    return ret;
+}
+
+/* ExtensionObject */
+CALC_JSON_TYPE(ExtensionObject) {
+    if(!src){
+        return calcWriteNull(ctx);
+    }
+    
+    u8 encoding = (u8) src->encoding;
+    if(encoding == UA_EXTENSIONOBJECT_ENCODED_NOBODY){
+        return calcWriteNull(ctx);
+    }
+    
+    status ret = UA_STATUSCODE_GOOD;
+    /* already encoded content.*/
+    if (encoding <= UA_EXTENSIONOBJECT_ENCODED_XML) {
+        UA_Boolean commaNeeded = UA_FALSE;
+        ret |= ADDJSONCHAR(ObjStart);
+
+        if(useReversible){
+            ret |= calcWriteKey(ctx, "TypeId", commaNeeded);
+            commaNeeded = true;
+            ret |= CALC_DIRECT(&src->content.encoded.typeId, NodeId);
+            if (ret != UA_STATUSCODE_GOOD)
+                return ret;
+        }
+        
+        switch (src->encoding) {
+            case UA_EXTENSIONOBJECT_ENCODED_BYTESTRING:
+            {
+                if(useReversible){
+                    UA_Byte jsonEncodingField = 1;
+                    ret |= calcWriteKey(ctx, "Encoding", commaNeeded);
+                    commaNeeded = true;
+                    ret |= CALC_DIRECT(&jsonEncodingField, Byte);
+                }
+                ret |= calcWriteKey(ctx, "Body", commaNeeded);
+                ret |= CALC_DIRECT(&src->content.encoded.body, String);
+                break;
+            }
+            case UA_EXTENSIONOBJECT_ENCODED_XML:
+            {
+                if(useReversible){
+                    UA_Byte jsonEncodingField = 2;
+                    ret |= calcWriteKey(ctx, "Encoding", commaNeeded);
+                    commaNeeded = true;
+                    ret |= CALC_DIRECT(&jsonEncodingField, Byte);
+                }
+                ret |= calcWriteKey(ctx, "Body", commaNeeded);
+                ret |= CALC_DIRECT(&src->content.encoded.body, String);
+                break;
+            }
+            default:
+                ret = UA_STATUSCODE_BADINTERNALERROR;
+        }
+
+        ret |= ADDJSONCHAR(ObjEnd);
```

Comment:
There is a huge if/else in this function. Its difficult to read for an outsider.
After the if/else block we return directly. So this can be simplified.

1. If the `if` is not visible from the `else` without scrolling, use comments to spell out what the else means
```c
} else { /* Now begins the case where... */
```
2. If the method returns directly after the if/else block, return from the function at the end of the if-block. And don't use an else at all.
3. If argument 2. is the case, put the less likely code path in the if-block. And the hot-path directly.
4. If large blocks of code remain, refactor into smaller methods called from a "master" function.
Use UA_assert to document what we know for sure about the input.

Reply:
This general hint can/should be applied in several similar places in this file.